### PR TITLE
add-ShoppingCart-model-firstCommit

### PR DESCRIPTION
**Pull request by OlegShvets1**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    private final CartItemMapper cartItemMapper;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartDto findShoppingCart(Authentication authentication, Pageable pageable) {
+        User user = (User) authentication.getPrincipal();
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Unable to locate a shopping cart with the given userId:  "
+                        + user.getId() + "Add items to the shopping cart to initialize it."));
```
Comment (olekskov) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]: I don't like idea when shopping cart can be absent when user exists.
Please create shopping cart for user during registration.
Service should assume that cart already exists if user is and handle it's absense only in minimum amount that may be connected to some technical problems with DB and like that

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        shoppingCartRepository.save(shoppingCart);
+        return cartItemMapper.toDto(cartItem);
+    }
+
+    @Override
+    public ShoppingCart createNewShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        return shoppingCartRepository.save(shoppingCart);
+    }
```
Comment (olekskov) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]: this logic -> during registration

**Review by olekskov — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -17,6 +19,7 @@ public class UserServiceImpl implements UserService {
     private final UserRepository userRepository;
     private final UserMapper userMapper;
     private final PasswordEncoder passwordEncoder;
+    private final ShoppingCartRepository shoppingCartRepository;
 
     @Override
     public UserResponseDto register(UserRegistrationRequestDto requestDto) {
```
Comment (olekskov) [inline src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java:27]: We have 2 operations in method that modify DB state, better to use relevant annotation
```suggestion
    @Transactional
    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
```

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    @Override
+    public ShoppingCartDto findShoppingCart(Authentication authentication, Pageable pageable) {
+        User user = (User) authentication.getPrincipal();
+
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> {
+                    ShoppingCart newShoppingCart = new ShoppingCart();
+                    newShoppingCart.setUser(user);
+                    return shoppingCartRepository.save(newShoppingCart);
+                });
```
Comment (olekskov) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]: I still don't like idea where we put something in DB during GET operation flow
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
                .orElseThrow(() -> new IllegalStateException("Cart cannot be found for..."));
```

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+        return shoppingCartDto;
+    }
+
+    @Override
+    public CartItemResponseDto addBookToShoppingCart(Authentication authentication,
+                                                     CartItemRequestDto requestDto) {
+        User user = (User) authentication.getPrincipal();
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> createNewShoppingCart(user));
```
Comment (olekskov) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
                .orElseThrow(() -> new IllegalStateException("Cart cannot be found for..."));
```

File: src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        shoppingCartRepository.save(shoppingCart);
+        return cartItemMapper.toDto(cartItem);
+    }
+
+    @Override
+    public ShoppingCart createNewShoppingCart(User user) {
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(user);
+        return shoppingCartRepository.save(shoppingCart);
+    }
```
Comment (olekskov) [inline src/main/java/mate/academy/bookstore/service/impl/ShoppingCartServiceImpl.java]:
```suggestion
```

**Review by olekskov — CHANGES_REQUESTED** [review]

**Review by kozhukhovsky — APPROVED** [review]

**Pull request merged**